[TEXT-212] A More Efficient Implementation for Calculating Size of Longest Common Subsequence

ali-ghanbari opened the pull request:
Existing implementation for the method `apply` of the class `LongestCommonSubsequence` calls the method `longestCommonSubsequence` of the class which is quadratic in time and space and it involves unnecessary string manipulation operations.
**Since the method `apply` is intended to calculate the size of Longest Common Subsequence (LCS), we do not need the whole 2D dynamic programming table nor do we need to calculate the actual LCS to take its length**.

This pull request proposes a direct solution for the problem of finding the size of LCS:
The idea is that if we are only concerned with calculating the size of LCS, we can implement a space efficient dynamic programming solution where we keep track of only the last two rows.
Please see [url](https://www.ics.uci.edu/~eppstein/161/960229.html) for more information.

Thank you.

kinow reviewed (commented):
Hi @ali-ghanbari thanks a lot for your PR and for the link with more details about the approach used.

I should have time in the next days to read the linked document and your code, confirm our tests have covered it and the result is the same, and merge it. And by coincidence I was planning on releasing a new version of Text in the next ~3 weeks. So perfect timing.

Thanks!

ali-ghanbari commented:
Thanks  @kinow 
Excellent!

XenoAmess commented:
Hi.
Thanks for your pr.
The thought is correct, it can save memory from n*m to 2* Math.min(n,m) in this way.
As you refined this function, you can refine **longestCommonSubsequence** as well.
use 3 numbers(a start-index and an end-index, and the length) to cache the dp
Of course you can also not cache the length but use start-index and end-index to calculate the length dynamically, but it need a jmh performance test to see whether way is better.

XenoAmess commented:
There also be some very small problems.
But I want you run some performance tests to see if it worthy to refine.
see the review.

XenoAmess commented:
Also, this pr worth a jira ticket.
https://issues.apache.org/jira/projects/IO/issues/

XenoAmess reviewed (changes requested):
notice that all of these are suggestions, which should run a jmh test to figure out whether worthy to apply

ali-ghanbari commented:
Hi @XenoAmess , @aherbert ,
Thank you for all your suggestions and comments.
I will look into them and update you in the up coming days.

ali-ghanbari commented:
Hi,

Thanks a lot for the comments! I just applied the suggestions to the previous implementation and also refactored it, so that it is now matching the algorithm description of in the original paper.

I also implemented linear space complexity algorithm for calculation of LCS.

ali-ghanbari commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -58,7 +58,48 @@ public Integer apply(final CharSequence left, final CharSequence right) {
+    static int calculateLCSLength(final CharSequence left, final int m,
+                                  final CharSequence right, final int n) {
+        // We only need to keep track of last two rows
+        final int[][] dp = new int[2][1 + n];
```
Please see the new implementation which matches the original paper. As for this comment, how about returning a pre-allocated singleton array with a 0 in it? The array won't escape this class (and package), so we can rest assured that its contents will always remain 0.

kinow replied:
>As for this comment, how about returning a pre-allocated singleton array with a 0 in it?

Sounds good to me :+1: worth adding a comment to the array so we don't accidentally make it public.

kinow replied:
@ali-ghanbari I think the paper uses `K`, any reason why `dp` instead of `k` (I guess `K` would upset Java linters)?

kinow commented:
Finished reviewing another PR for Commons Text, now this one is the next in the list. Sorry for the long wait @ali-ghanbari

ali-ghanbari commented:
> Finished reviewing another PR for Commons Text, now this one is the next in the list. Sorry for the long wait @ali-ghanbari

Glad to hear that @kinow ! Many thanks.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -58,7 +58,57 @@ public Integer apply(final CharSequence left, final CharSequence right) {
+    }
+
+    /**
+     * An implementation of "ALG B" from Hirschberg's paper <a href="https://dl.acm.org/doi/10.1145/360825.360861">A linear space algorithm for computing maximal common subsequences</a>.
```
This reference is fine here, but I think it's worth either moving or duplicating in the main comment for this class, please. There's another reference there already.

ali-ghanbari replied:
> ... a comment to the array so we don't accidentally make it public.

How about checking the input sizes before invoking althorithmB? As @kinow mentioned, in that case we have to make sure that the method is not public so as to avoid malfunctioning of the method. I think it is a good idea to leave the implementation as it is and just check the input sizes before calling the method. Please see the newest implementation and share your ideas.

> I think the paper uses K, any reason why dp instead of k (I guess K would upset Java linters)

Yes, the paper uses notation that might be considered problematic for this code-base. To avoid confusion, I renamed the variable to "dpRows," which is what it is (it is two rows of the DP table in the traditional algorithm).

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -58,7 +58,57 @@ public Integer apply(final CharSequence left, final CharSequence right) {
+    /**
+     * An implementation of "ALG B" from Hirschberg's paper <a href="https://dl.acm.org/doi/10.1145/360825.360861">A linear space algorithm for computing maximal common subsequences</a>.
+     * Assuming the sequence <code>left</code> is of size <code>m</code> and the sequence <code>right</code> is of size <code>n</code>,
+     * this method returns the last row of the dynamic programming table when calculating LCS the two sequences.
```
s/calculating LCS the two sequences/calculating the LCS of the two sequences

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -58,7 +58,57 @@ public Integer apply(final CharSequence left, final CharSequence right) {
+     * this method returns the last row of the dynamic programming table when calculating LCS the two sequences.
+     * Therefore, the last element of the returned array, is the size of LCS of <code>left</code> and <code>right</code>.
+     * This method runs in O(m * n) time and O(n) space.
+     * To save more space, it is preferable to pass the shorter sequence as <code>right</code>.
```
> To save more space, it is preferable to pass the shorter sequence as <code>right</code>.

Worth having a pre-step that checks the length and swaps left/right if needed? Quite sure we have this happening somewhere else in Commons Text already.

ali-ghanbari replied:
Yes, that would be more _user-friendly_, but as algorithmB is being called from algorithmC as an auxiliary function and swapping the parameters causes the method to return an array of some length that is not expected by algorithmC. That's why I didn't adopt that approach.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -120,24 +170,72 @@ public CharSequence longestCommonSubsequence(final CharSequence left, final Char
+     * @param n Length of right sequence
+     * @return LCS of <code>left</code> and <code>right</code>
+     */
+   static CharSequence algorithmC(final CharSequence left, final int m,
```
@ali-ghanbari if I understood correctly, Algorithm A and Algorithm B are very similar, but I think A requires `O(mn)` time and `O(mn)` space. While Algorithm B requires `O(mn)` time and `O(m+n)` space.

And, Algorithm C is the third one in that paper, with `O(mn)` time as Algorithms A and B, and `O(m+n)` space. With the difference between B and C being that B uses temporary memory linear to `m` and `n`. Whereas Algorithm C uses a constant amount of memory.

(Past midnight here, long time with no :coffee: so forgive me if I missed something, or said something wrong above :+1: )

The `LongestCommonSubsequence` is a function supposed to be called as a `BiFunction`, with `apply()` being the main method, and entry point to the algorithm.

With the changes here, I think now `apply` uses Algorithm B. While this method with Algorithm C is called only from the public `longestCommonSubsequence`.

Is there a reason for having both implementations here, instead of just Algorithm C and calling it from the `apply` method?

ali-ghanbari replied:
>  ... Algorithm A and Algorithm B are very similar, but I think A requires O(mn) time and O(mn) space. While Algorithm B requires O(mn) time and O(m+n) space.

Right, in terms of time complexity, yes, they all are quadratic (O(mn), to be precise) in the size of input strings.
Algorithm B has O(m+n) space complexity, while Algorithm A has O(mn) space complexity (it returns an entire DP array; just like the method that we marked as deprecated).

> And, Algorithm C is the third one in that paper, with O(mn) time...

Thanks for bring this up @kinow. The method algorithmC allocates heap space directly or indirectly (through the calls to subSequence), and there are 2m-1 calls to algorithmC. The recurrence does not seem to be easy to solve for me, but a back of the envelope calculation tells me that this implementation should not be of linear space complexity...
Looking at the space analysis proof given in the paper, it looks like the paper makes a simplifying assumption about the storage of substrings. (please correct me if I am wrong)
Perhaps that is why [https://www.ics.uci.edu/~eppstein/161/960229.html](link), emphasizes that Hirschberg's linear-spaced algorithm is used **only** for calculating length of the LCS.
Please confirm this and I can delete algorithmC and keep algorithmB only.
I can quickly adjust the rest of the class code accordingly. Thanks!

> Past midnight here, long time with ...

Thanks @kinow your constructive comments and your hard work! 👍 🥇

kinow replied:
Hi @ali-ghanbari !

We are almost there. Just wanted to check a few more things before digging into algorithm C and finish reading the paper and code here.

> Algorithm C calculates the LCS itself (not just its length). If the class LongestCommonSubsequence is not supposed to calculate the LCS of two strings, then there is no reason to keep Algorithm C.

Agree. Since the method to calculate the longest subsequence was public, we will have to maintain it, and will have to calculate the LCS itself. In this case, I agree we should keep the Algorithm C, since it has the best space + time.

>To implement apply method, we could use Algorithm C, as well. We could call algorithmC and take the length of the returned CharSequence and return it, but that would involve extra computations; Algorithm B can be directly used.

Right. But we would be keeping both algorithms. I think the optimization we gain from using Algorithm B and Algorithm C, is that — if I understood correctly — the `apply` method will call Algorithm B, which may use more space/memory, but won't allocate an Object for the LCS itself, instead only returning the length from the computation array/table.

On the other hand, we could keep just one algorithm, allocate the strings, and use it for both the LCS method, and for the `apply` method to return the similarity score. I **think** I prefer to keep a single algorithm for the sake of maintainability. And since C performs better, I'd keep just C.

And maybe add a note to the Javadoc that it would be possible to optimize the `apply` by using the Algorithm B of the paper (or maybe someone could come up with a way to split the Algorithm C in a way that returns the length or the LCS). But this way we would have a good improvement over the current status quo, and introduce code that is not hard to maintain.

If in the future users want to further optimize, then we have the option of adding Algorithm B back.

What do you think?

ali-ghanbari replied:
Hello!

> We are almost there. Just wanted to check a few more things before digging into algorithm C and finish reading the paper and code here.

I just wanted to let you know that I have made some refactoring.
Instead of just passing string sequences to `algorithmB` and `algorithmC`, I have modified the methods to take the strings with their start and end index.
I also added an extra parameter to `algorithmB` so that we can traverse the string backward instead of literally reversing the input strings (in previous version of `algorithmC` we were reversing the arguments in one of the calls to `algorithmB`).
This refactoring should certainly increase the _dexterity_ of the implementation as we avoid several object creation, string reversal, and virtual method calls (a JMH might be needed anyway, as other maintainers also mentioned).
Besides that, I think this refactor was, in fact, necessary to make the implementation as similar to the original algorithm as possible: the proof for space analysis of Algorithm C says "_We assume [...] and substrings can be transferred as arguments by giving initial and final locations_."
I hope with this change, I have not interrupted your review process! Please let me know if adding more comments helps.

> I think I prefer to keep a single algorithm for the sake of maintainability. And since C performs better, I'd keep just C.
> What do you think?

I think we cannot get rid of Algorithm B, as it is being used as an auxiliary procedure in implementing Algorithm C: please see step 3 of Algorithm C in the paper.
But, I understand the maintainability concern. 
How about doing this: we could keep **only** Algorithm B and in this version to improve `apply` method only; we can keep the old implementation (dynamic programming) of the method `longestCommonSubsequence` for now.
We can mention that improving space complexity of `longestCommonSubsequence` is a work in progress.

But if we want to keep Algorithm C, we must keep Algorithm B as it is being called from Algorithm C.
And in that case it is ideal to use Algorithm B for `apply` method, because we already have the implementation and it involves less computation than Algorithm C.

Please let me know if you prefer the former approach. In that case I can revert the code for `longestCommonSubsequence`.

Thanks a lot!

kinow reviewed (commented):
@ali-ghanbari thanks again for the great PR. I was surprised to see the paper is from the 70's! I think the last time I searched for LCS in GitHub (in Java/Python) the implementations were similar to algorithm A. Good improvement for Commons Text.

I read the paper up to the algorithm C (need to finish reading it) comparing the paper with your code :+1: 

Take a look at the comments in this review, and after you have addressed them, I will go once again, now with the debugger just to clear a few things and make sure I understand the improvements in the paper (really interested to grok it :neckbeard: ), and to finalize reviewing your `algorithmC` method.

Great work, kudos!

ali-ghanbari commented:
> @ali-ghanbari thanks again for the great PR. I was surprised to see the paper is from the 70's! I think the last time I searched for LCS in GitHub (in Java/Python) the implementations were similar to algorithm A. Good improvement for Commons Text.

Thanks @kinow :) Yes, the idea has been around for quite a long time; that was surprising to me as well when I first learned about this newer algorithm.
Right, traditional implementations are more or less the same as "Algorithm A," according to that CACM paper (which is then refined by Algorithm B/C in the paper).

> ... I will go once again, now with the debugger just to clear a few things and make sure I understand the improvements in the paper ...

Sounds good! Thanks!

ali-ghanbari commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -120,24 +170,72 @@ public CharSequence longestCommonSubsequence(final CharSequence left, final Char
            throw new IllegalArgumentException("Inputs must not be null");
        }
-       final StringBuilder longestCommonSubstringArray = new StringBuilder(Math.max(left.length(), right.length()));
-       final int[][] lcsLengthArray = longestCommonSubstringLengthArray(left, right);
```
I marked the method as deprecated.

ali-ghanbari commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -58,7 +58,57 @@ public Integer apply(final CharSequence left, final CharSequence right) {
+     * @param n Length of right sequence
+     * @return Last row of DP table for calculating LCS of <code>left</code> and <code>right</code>
+     */
+    static int[] algorithmB(final CharSequence left, final int m,
```
Yes, that's the idea! ;)
The traditional LCS algorithms are variants of Algorithm A in the paper.

garydgregory commented:
We'll need some JMH benchmarks to show what code performs best before picking a new method IMO...

ali-ghanbari commented:
> We'll need some JMH benchmarks to show what code performs best before picking a new method IMO... otherwise this is a no-go from my POV, it's just changing code...

Agreed.
Please note that this implementation improves **space complexity** (from quadratic to linear, which is a huge improvement!) and the general speed of the methods (asymptotically speaking) should remain more or less the same (as both are quadratic in time). I am not sure if JMH can measure memory footprint, but to test the memory usage, I suggest testing the new implementation with the first pair of large strings that make the previous implementation crash in one way or another.
I happened to learn about this more efficient algorithm in a research project where I was getting OutOfMemoryError with the current implementation of `apply` method and I used this new implementation and it worked without any problem.

kinow commented:
@ali-ghanbari

>But if we want to keep Algorithm C, we must keep Algorithm B as it is being called from Algorithm C.
And in that case (in my opinion) it is ideal to use Algorithm B for apply method, because we already have the implementation and it involves less computation than Algorithm C.

You convinced me that the current approach of Algorithm B + Algorithm C is the best option :-), I missed the part that C was using B, sorry! And thanks for understanding and pointing it so politely.

>I happened to learn about this more efficient algorithm in a research project where I was getting OutOfMemoryError with the current implementation of apply method and I used this new implementation and it worked without any problem.

Interesting! I think JMH should be able to show that the new code performs better memory-wise too.

@ali-ghanbari I will finish reading the paper either this weekend or next week, sync your branch once more, and then go one last time through the code. If I don't find anything missing, except for the missing JMH test (which I could try to help if you don't have time, it's been ages since I wrote a JMH benchmark, good to practice) it should be ready to be approved!

Thanks a lot.
Bruno

ali-ghanbari commented:
Sounds good @kinow :) Thanks a lot!

I will update you as soon as I manage to run the JMH tests.

kinow reviewed (changes requested):
I think CI was passing before, but apparently some Javadoc needs updating @ali-ghanbari ?

```
 Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.2.0:javadoc (default-cli) on project commons-text: An error has occurred in Javadoc report generation: 
Error:  Exit code: 1 - /home/runner/work/commons-text/commons-text/src/main/java/org/apache/commons/text/StrTokenizer.java:181: error: unexpected end tag: </p>
Error:       * </p>
Error:         ^
```

ali-ghanbari commented:
> I think CI was passing before, but apparently some Javadoc needs updating @ali-ghanbari ?
> 
> ```
>  Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.2.0:javadoc (default-cli) on project commons-text: An error has occurred in Javadoc report generation: 
> Error:  Exit code: 1 - /home/runner/work/commons-text/commons-text/src/main/java/org/apache/commons/text/StrTokenizer.java:181: error: unexpected end tag: </p>
> Error:       * </p>
> Error:         ^
> ```

Yes, @kinow . I have just rectified the issue. I hope this didn't interrupt your review.

ali-ghanbari commented:
Thank you @kinow for mentioning the CI issue.

It looks like that the issue is due to a recent change([see](https://github.com/apache/commons-text/commit/87640ed259da32f6af516b9b638ae46d96c9ff24) for the update that resolves the CI errors) in CI configuration which makes the older version of `StrTokenizer.java` (which contains several dangling `p` tags in its Javadocs) to fail checkstyle checks.
I guess the issue can be ignored for the time being? Please let me know if we should update `StrTokenizer.java` according to the latest version.

Please also see the newest version of the implementation which avoids checkstyle warnings by passing the arguments to `algorithmB` using instance fields (it used to complain because we had methods with more than 7 parameters).
With this change it is imperative to have _separate instances_ of the class in case if we want to run the algorithm concurrently.

By the way, I managed to run JMH. Please check out the Javadoc for `apply` method to see the results.

Thank you again!

kinow commented:
@ali-ghanbari , I will review it once I have some spare time. Thanks a lot for updating the PR and fixing CI issues.

We can definitely ignore that issue since it is not being introduced here (will fix it in a bit). As for the JMH test, we could probably include it in the source code (if not already in this PR) and run only when a certain Maven profile is activated, or filter the suite with the test, etc.

Thanks
Bruno

ali-ghanbari commented:
> @ali-ghanbari , I will review it once I have some spare time. Thanks a lot for updating the PR and fixing CI issues.
> 
> We can definitely ignore that issue since it is not being introduced here (will fix it in a bit). As for the JMH test, we could probably include it in the source code (if not already in this PR) and run only when a certain Maven profile is activated, or filter the suite with the test, etc.
> 
> Thanks
> Bruno

Thanks a lot!
FYI, for now JMH is in a separate project.
I will figure it out how to embed the test code in this project (without having name conflicts) and selectively run it using Maven.

kinow commented:
There is already a JMH test somewhere in the code, added recently to Text (I was looking at Numbers & Imaging, as I recalled we added one somewhere really recently -  turns out it was here after all :grimacing: ).

https://github.com/apache/commons-text/blob/78a818538869e51492c4a2bebd42322acceb660b/pom.xml#L537-L575

Just have a look at the Maven profile & at the existing benchmark code. It should hopefully be easy to include your JMH test there and run it in that Maven profile too.

kinow commented:
I think I fixed the Javadoc error in `master`. If you could rebase it later, then CI should be OK again. BTW, when we are ready to merge it, if I am the one doing the merge, I probably won't be using GH UI, but local command line. My first step will be to squash all 20 commits or more into a single commit.

If you'd like to do that yourself, or if you think it would be appropriate to keep some separate commits (e.g. some of the changes done during review could be undone later, or the commits can explain a certain decision), then feel free to squash it before the final review :+1: 

Thanks @ali-ghanbari !

ali-ghanbari commented:
Hello,

Thank you @kinow for the insightful comment regarding `StringBuilder` object used in `algorithmC`.
It took me a while to understand that it is possible to allocate the object outside of the method, but I finally managed to code it!
I think `algorithmC` is now satisfying all the assumptions made by the paper and a back of the envelope calculation will reveal that the method has indeed linear space complexity.

I have made a whole bunch of refactoring and added lots of comments and Javadoc to improve readability of the code.
I have also deleted the method logestCommonSubsequence for you (the one that had typo in its name and it was planned to be removed in version 2.0).

From now on, unless the reviewers request some changes, I am unlikely to make updates to the class. So, I would say, it is safe to review :)

I will add the code for JMH as soon as I find some spare time.

Thanks a lot!

kinow commented:
@ali-ghanbari we are not releqsing 2.0 yet. That's why the method with a typo was not removed. Could you put it back, please?

kinow commented:
Do you mind squashing the commits too? Adding an extra commit to add JMH should be fine and doesn't need to be squashed if you don't want to, but if the next review doesn't find any errors, we won't need to touch the branch. just approve and merge, which is a bit better (we could accidentally miss a commit when squashing for instance).

Thanks

ali-ghanbari commented:
> @ali-ghanbari we are not releqsing 2.0 yet. That's why the method with a typo was not removed. Could you put it back, please?

Sorry for misunderstanding! I returned the method back :)

ali-ghanbari commented:
> Do you mind squashing the commits too? Adding an extra commit to add JMH should be fine and doesn't need to be squashed if you don't want to, but if the next review doesn't find any errors, we won't need to touch the branch. just approve and merge, which is a bit better (we could accidentally miss a commit when squashing for instance).
> 
> Thanks

Not at all! I have just squashed the commits.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -91,65 +309,182 @@ public CharSequence logestCommonSubsequence(final CharSequence left, final CharS
      * O(m*n), where m=left.length() and n=right.length().
      *
-     * @param left first character sequence
+     * @param left  first character sequence
```
Unnecessary space.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -91,65 +309,182 @@ public CharSequence logestCommonSubsequence(final CharSequence left, final CharS
+     * @param left  first character sequence
      * @param right second character sequence
      * @return lcsLengthArray
+     * @deprecated Deprecated as of 2.0. A more efficient implementation for calculating LCS is now available.
```
Not 2.0. Deprecated as of 1.10... And add a note that this method will be removed in 2.0. Similar to `logestCommonSubsequence`'s comments.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * These arrays are used as buffers for the calls to {@link #algorithmB(int[][], boolean)}, as per the paper.
+     * The contents of the arrays shall be ignored and their contents will be modified.
+     *
+     * @since 2.0
```
Not since 2.0. Either leave blank or @since 1.10.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * An object of <code>StringBuilder</code> is expected.
+     * The method shall store its results in this field.
+     *
+     * @since 2.0
```
Not since 2.0. Either leave blank or @since 1.10.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * We are passing some of the parameters through fields to avoid defining methods with more than 7 parameters
+     * which makes checkstyle plugin to complain.
+     *
+     * @since 2.0
```
Not since 2.0. Either leave blank or @since 1.10 (which is the next release according to the `pom.xml` version, which is 1.10.0-SNAPSHOT (i.e. the snapshot part is dropped, and we just cut a new release; eventually there will come a time we will be ready for 2.0, but I don't think we are planning it anytime soon yet)

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * A parameter of {@link #algorithmB(int[][], boolean)} and {@link #algorithmC(int, int, int, int)}:
+     * The first input sequence.
+     * We are passing some of the parameters through fields to avoid defining methods with more than 7 parameters
+     * which makes checkstyle plugin to complain.
```
I think this is not the best option. It's dangerous to maintain state in the object just to pass checkstyle tests @ali-ghanbari. It would be much safer to pass the parameters around, without ever maintaining state in the class (I think that's how the old implementation was done).

Can you use the old approach (as you had before), please? If Checkstyle fails, we can just add a suppression to `checkstyle-suppressions.xml` in the root directory of the project.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * We are passing some of the parameters through fields to avoid defining methods with more than 7 parameters
+     * which makes checkstyle plugin to complain.
+     *
+     * @since 2.0
```
Not since 2.0, either leave blank or @since 1.10

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * We are passing some of the parameters through fields to avoid defining methods with more than 7 parameters
+     * which makes checkstyle plugin to complain.
+     *
+     * @since 2.0
```
Not since 2.0, either leave blank or @since 1.10

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * A parameter of {@link #algorithmB(int[][], boolean)} and {@link #algorithmC(int, int, int, int)}:
+     * The second input sequence.
+     *
+     * @since 2.0
```
Not since 2.0, either leave blank or @since 1.10

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * We are passing some of the parameters through fields to avoid defining methods with more than 7 parameters
+     * which makes checkstyle plugin to complain.
+     *
+     * @since 2.0
```
Not since 2.0, either leave blank or @since 1.10

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * We are passing some of the parameters through fields to avoid defining methods with more than 7 parameters
+     * which makes checkstyle plugin to complain.
+     *
+     * @since 2.0
```
Not since 2.0, either leave blank or @since 1.10

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,254 @@
+     * @param reverseSequences If true {@code left} and {@code right} will be accessed in reverse order
+     * @return Last row of DP table for calculating the LCS of <code>left</code> and <code>right</code>
+     *
+     * @since 2.0
```
Not since 2.0, either leave blank or @since 1.10

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -91,65 +309,182 @@ public CharSequence logestCommonSubsequence(final CharSequence left, final CharS
+     * @param rightStart Start index of the first input subsequence in {@code right}
+     * @param rightEnd   End index (inclusive) of the first input subsequence in {@code right}
      *
+     * @since 2.0
```
Not since 2.0, either leave blank or @since 1.10

kinow reviewed (changes requested):
Hi @ali-ghanbari 

I was going to start reviewing algorithm C, but I realized you had a few more differences in the pull request from the last time I read your code.

There are problems with the current implementation, namely the @since version that is not correct, one unnecessary blank space (I can fix that later if needed when merging), and the most important that was the parameters for the function that - I think - you moved to members of the class so that Checkstyle would not complain.

It would be better to avoid maintaining state in the object if that's not necessary (someone reading the code will immediately know it's thread-safe, since it has no dependency to other classes, and no local state, among other benefits).

Your previous version was really good, very close to being ready to be merged. Could you undo these changes I mentioned above, keeping whatever you had about algo B and C, please?

Thanks!
Bruno

ali-ghanbari commented:
> Hi @ali-ghanbari
> 
> I was going to start reviewing algorithm C, but I realized you had a few more differences in the pull request from the last time I read your code.
> 
> There are problems with the current implementation, namely the @SInCE version that is not correct, one unnecessary blank space (I can fix that later if needed when merging), and the most important that was the parameters for the function that - I think - you moved to members of the class so that Checkstyle would not complain.
> 
> It would be better to avoid maintaining state in the object if that's not necessary (someone reading the code will immediately know it's thread-safe, since it has no dependency to other classes, and no local state, among other benefits).
> 
> Your previous version was really good, very close to being ready to be merged. Could you undo these changes I mentioned above, keeping whatever you had about algo B and C, please?
> 
> Thanks!
> Bruno

Certainly, @kinow :) I will address all of your comments soon.
By the way, JMH Maven profile and the test code are ready; I can push that also.

Thanks!

kinow commented:
> By the way, JMH Maven profile and the test code are ready; I can push that also.

Brilliant! Thanks heaps for the super fast reply @ali-ghanbari !

ali-ghanbari commented:
Hello @kinow ,

Thank you for all your constructive comments!

I have reverted the code to its older form, thereby sacrificing (slight gain in) preformance to readability of the code.

I have also added JMH performance analyzer.

Side note: I have not forgotted your comments about whitespaces; it looks like my editor adds them automatically when I reformatted the code to adjust the indentation. I can rectify them if you don't mind another commit.

Thank you,
Ali

kinow commented:
>Side note: I did not foget your comments about whitespaces; it looks like my editor adds them automatically when I reformatted the code to adjust the indentation. I can rectify them if you don't mind another commit.

IntelliJ? I use both Eclipse and IntelliJ, and sometimes I find IntelliJ does a few things like this, especially after an update. I don't mind reverting that when it's ready to merge (takes just a couple minutes). So no need to bother about this :+1: 

@ali-ghanbari GitHub UI is showing pom.xml as conflicting now.

ali-ghanbari commented:
> IntelliJ? ... I don't mind reverting that when it's ready to merge ...

Yes, I mostly use IntelliJ.
Resolving `pom.xml` conflicts turned out to be an opportunity for me to fix the whitespace problems ;)

kinow commented:
Thanks @ali-ghanbari I've approved the GH Actions job, and it should finish in a few minutes :crossed_fingers: I should be able to review it again this weekend.

kinow commented on pom.xml:
```diff
@@ -503,5 +535,44 @@
+            <version>3.0.0</version>
+            <executions>
+              <execution>
+                <id>benchmark</id>
```
Strange, we already have a `benchmark` Maven profile. Did you rebase when you added your JMH changes @ali-ghanbari ? https://github.com/apache/commons-text/blob/c4b2c2774b3d797df74a87418c4818aa6d0a8538/pom.xml#L538

ali-ghanbari replied:
Yes @kinow, we do have a JMH benchmark profile already in the latest version in `master` branch.
I actually copied from there to reduce the risk of conflicts.

> Did you rebase when you added your JMH changes?

I just manually copied relevant parts into `pom.xml`.

ali-ghanbari replied:
@kinow :
And since we already have a JMH benchmark, I suggest keeping LCS's benchmark within a separate package. Is that necessary?

kinow commented on src/test/java/org/apache/commons/text/jmh/LongestCommonSubsequencePerformance.java:
```diff
@@ -0,0 +1,165 @@
+            final ClassLoader classloader = Thread.currentThread().getContextClassLoader();
+            try (final InputStream is = classloader.getResourceAsStream("lcs-perf-analysis-inputs.csv");
+                 final InputStreamReader isr = new InputStreamReader(Objects.requireNonNull(is));
+                 final BufferedReader br = new BufferedReader(isr)) {
```
Travis CI is failing due to these `final` statements above, but no need to do anything about it. We can turn that off in Travis/checkstyle :+1:

ali-ghanbari replied:
> ... but no need to do anything about it
It's your call. If we decide to put LCS's benchmark into a separate folder, I can rectify that right quick ;)

kinow commented:
Rebased PR onto latest `master`. Diff in GH UI now showing only code added for this PR. Also amending commit now to include the JIRA ticket.

kinow commented:
Placeholder ticket: https://issues.apache.org/jira/browse/TEXT-212

kinow commented on src/test/resources/lcs-perf-analysis-inputs.csv:
```diff
@@ -0,0 +1,5 @@
+"You should have received a copy of the GNU General Public License version","2 along with this work; if not, write to the Free Software Foundation,"
+"Here, the field iterations will be populated with appropriate values from the @Param annotation by the JMH when it is passed to the benchmark method. The @Setup annotated method is invoked before each invocation of the benchmark and creates a new Hasher ensuring isolation. When the execution is finished, we'll get a result similar to the one below: When running microbenchmarks, it's very importa ...
+"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa q ...
+"But I must explain to you how all this mistaken idea of denouncing pleasure and praising pain was born and I will give you a complete account of the system, and expound the actual teachings of the great explorer of the truth, the master-builder of human happiness. No one rejects, dislikes, or avoids pleasure itself, because it is pleasure, but because those who do not know how to pursue pleasure ...
```
Added missing new line (squashed change-commit)

kinow reviewed (commented):
Confirmed time/space complexity of the new algorithm B looks good. Had a quick look at the code since it changed a bit from the last I looked at it. Pushed a few changes, squashing commits, but only cosmetic to CI/linters are happy.

Now onto final algorithm C and hopefully push the approve button :drum:

kinow commented on src/test/java/org/apache/commons/text/jmh/LongestCommonSubsequencePerformance.java:
```diff
@@ -0,0 +1,166 @@
+                 BufferedReader br = new BufferedReader(isr)) {
+                String line;
+                while ((line = br.readLine()) != null && !line.trim().isEmpty()) {
+                    line = line.trim();
```
Checkstyle was complaining about re-assigning `line` in the `while` block, so I left the double `line.trim` call, which shouldn't really interfere with the performance tests :+1:

ali-ghanbari replied:
I appreciate it.

kinow commented on src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java:
```diff
@@ -29,36 +29,109 @@
+     * An implementation of "ALG B" from Hirschberg's CACM '71 paper.
+     * Assuming the first input sequence is of size <code>m</code> and the second input sequence is of size
+     * <code>n</code>, this method returns the last row of the dynamic programming (DP) table when calculating
+     * the LCS of the two sequences in <i>O(m*n)</i> time and <i>O(n)</i> space.
```
@ali-ghanbari 

>and <i>O(n)</i> space

Isn't that `O(m+n)`?

kinow replied:
Oh, I think because you are always passing the smallest as the second sequence, then it'll be `O(n)`?

kinow replied:
Although `algorithmC` does similar trick, but in the javadocs we say `O(m+n) space`

ali-ghanbari replied:
Hi @kinow , thank you so much for your review :)

>  ... I think because you are always passing the smallest as the second sequence, then it'll be O(n)?

Yes, that is because of the way we allocate the array.

> Although algorithmC does similar trick, but in the javadocs we say O(m+n) space

Yes, _on the heap_ it is O(n), but since the implementation is recursive, it goes O(m) deep in the stack also.

ali-ghanbari commented:
Thank you @kinow once again for your review, refactoring, and constructive feedback :+1: 

I checked the comments. Hopefully, I have answered all of them. Please kindly remind me in case I have missed any.

Ali

kinow commented:
Hi @ali-ghanbari 

Thanks for the prompt reply. I went out for lunch, and am back to whiteboard (literally a small whiteboard). Where I'm having fun after zooming-in as much as I could into the pseudocode of algorithm C. I was trying to remember if I ever learned the syntax `Â`. Is that a symbol for inverted `A`, or simply a mark to differentiate from `A`?

I am using a very simplified `right` and `left` for testing (`A='abc'` and `B='bc'`) just to have a look at what the tables/arrays returned look like, and compare with our code. As for time & space complexity, I believe you have the same results as the paper :clap: 

Once I manage to run the pseudocode of algorithm C in my whiteboard I should be finished with the review :-) thanks!

ali-ghanbari commented:
@kinow ,

> ... zooming-in as much as I could into the pseudocode of algorithm C.

I agree that the paper, being printed a long time ago, is not as clean as the papers we see today and it has lots of crooked letters that need zooming in ;)

> I was trying to remember if I ever learned the syntax `Â`. Is that a symbol for inverted `A`, or simply a mark to differentiate from `A`?

That's where zooming in helped me :D The notation is defined in the 4th paragraph of the second column on the first page (but it was difficult for me to notice!): it is meant to denote the reverse/inverse of the string. So, yes, `Â` is used by the author to denote inverted `A`.

> I am using a very simplified `right` and `left` for testing (`A='abc'` and `B='bc'`) just to have a look at what the tables/arrays returned look like, and compare with our code. As for time & space complexity, I believe you have the same results as the paper...
> Once I manage to run the pseudocode of algorithm C in my whiteboard I should be finished with the review

Sounds good!

Thanks,
Ali

kinow reviewed (approved):
Looks good! And to the best I could tell, algorithm C is performing as well as described in the paper (with little performance improvement for choosing the order of strings, swapping array elements in Java, etc).

Had fun following the pseudo-code with some simple strings to understand the algorithms :grimacing: 

![20210806_101218](https://user-images.githubusercontent.com/304786/128448868-68d1fcac-a474-406c-9865-313e9d7e4833.jpg)

![20210806_144233](https://user-images.githubusercontent.com/304786/128448874-736011f0-3a1f-4da3-be63-c452a0555e2a.jpg)

Then read the Java code trying to keep track of space & time, and looks like @ali-ghanbari has the improvements described in the paper.

Great job @ali-ghanbari ! Thanks a lot for your patience and for your contribution :+1: 

Bruno

ali-ghanbari commented:
Thanks for your review, @kinow ! And I appreciate your efforts in tracing the pseudo-code and creating more confidence in its correctness for those of us who didn't follow the correctness proof.

Ali

kinow commented:
@aherbert , @XenoAmess if there are no other comments/feedback, I'll merge this PR next week or week after :+1:

kinow commented:
Merged, JIRA issue updated. Thank you a lot for your contribution @ali-ghanbari !
